controllers/managedclustermodule_reconciler.go

```diff
@@ -0,0 +1,156 @@
+	return &ManagedClusterModuleReconciler{
+		Client:      client,
+		manifestAPI: manifestAPI,
+	}
+}
+
+func (r *ManagedClusterModuleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
+
+	logger := log.FromContext(ctx)
+
+	mcm, err := r.getRequestedManagedClusterModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("ManagedClusterModule deleted")
+			return res, nil
+		}
+
+		return res, fmt.Errorf("failed to get the requested CR: %v", err)
+	}
+
+	logger.Info("Requested KMMO ManagedClusterModule")
+
+	clusters, err := r.getSelectedManagedClusters(ctx, mcm)
+	if err != nil {
+		return res, fmt.Errorf("failed to get selected clusters: %v", err)
+	}
+
+	for _, cluster := range clusters.Items {
+		mw := &workv1.ManifestWork{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      mcm.Name,
+				Namespace: cluster.Name,
+			},
+		}
+
+		opRes, err := controllerutil.CreateOrPatch(ctx, r.Client, mw, func() error {
+			return r.manifestAPI.SetManifestWorkAsDesired(ctx, mw, *mcm)
+		})
+
+		if err != nil {
+			return res, fmt.Errorf("failed to create/patch ManifestWork for managed cluster %s: %v", cluster.Name, err)
+		}
+
+		logger.Info("Reconciled ManifestWork", "name", mw.Name, "namespace", mw.Namespace, "result", opRes)
+	}
+
+	if err := r.manifestAPI.GarbageCollect(ctx, *clusters, *mcm); err != nil {
+		return res, fmt.Errorf("failed to garbage collect ManifestWorks with no matching cluster selector: %v", err)
+	}
+
+	return res, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ManagedClusterModuleReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kmmv1beta1.ManagedClusterModule{}).
+		Owns(&workv1.ManifestWork{}).
+		Watches(
+			&source.Kind{Type: &clusterv1.ManagedCluster{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(r.getSkipUpdatesPredicate())).
+		Named("managedclustermodule").
+		Complete(r)
+}
+
+func (r *ManagedClusterModuleReconciler) getRequestedManagedClusterModule(
+	ctx context.Context,
+	namespacedName types.NamespacedName) (*kmmv1beta1.ManagedClusterModule, error) {
+
+	mcm := kmmv1beta1.ManagedClusterModule{}
+	if err := r.Client.Get(ctx, namespacedName, &mcm); err != nil {
+		return nil, fmt.Errorf("failed to get the ManagedClusterModule %s: %w", namespacedName, err)
+	}
+	return &mcm, nil
+}
+
+func (r *ManagedClusterModuleReconciler) getSelectedManagedClusters(
+	ctx context.Context,
+	mcm *kmmv1beta1.ManagedClusterModule) (*clusterv1.ManagedClusterList, error) {
+
+	clusterList := &clusterv1.ManagedClusterList{}
+
+	opts := []client.ListOption{
+		client.MatchingLabelsSelector{
+			Selector: labels.Set(mcm.Spec.Selector).AsSelector(),
+		},
+	}
+	err := r.Client.List(ctx, clusterList, opts...)
+
+	return clusterList, err
+}
+
+func (r *ManagedClusterModuleReconciler) getSkipUpdatesPredicate() predicate.Predicate {
```

Comment:
we have a filter package that contains all the filters for our reconcilers, i think it is better to move this function there

Reply:
Thanks for pointing this out. Given [this](https://github.com/kubernetes-sigs/kernel-module-management/pull/153#discussion_r1023775040) resolution, I have removed this method.